src/Security/InheritedPermissions.php

```diff
@@ -0,0 +1,632 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use InvalidArgumentException;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\Hierarchy\Hierarchy;
+use SilverStripe\Versioned\Versioned;
+
+/**
+ * Calculates batch permissions for nested objects for:
+ *  - canView: Supports 'Anyone' type
+ *  - canEdit
+ *  - canDelete: Includes special logic for ensuring parent objects can only be deleted if their children can
+ *    be deleted also.
+ */
+class InheritedPermissions
```

Comment:
Well, we have 
checkOne($permission, $id, $member)
checkMany($permission, $ids, $member)

and we have
check($permission, $member) which is self-scoped to a record. This covers both Datobjects themselves, and the RootPermissions implementors.

I'm having trouble thinking of names to distinguish these interfaces that makes their purpose clear.